### ISSUE TITLE
ENH: expose pygeos functions as GeoSeries/GeometryArray methods

### ISSUE THREAD

**Issue by martinfleis** [geopandas/geopandas#2010]
`pygeos` now contains some functions which could be exposed as methods on GeoSeries/GeometryArray levels. Some of them have shapely (1.7) versions, most do not.

I would not mind supporting them only if pygeos is installed and mark them as experimental until shapely 2.0 is out. The implementation of these should be relatively easy and straightforward then.

Probably incomplete list of candidates:

Measurement:
- [ ] frechet_distance
- [ ] hausdorff_distance
- [ ] minimum_bounding_radius
- [ ] minimum_clearance

Predicates:
- [ ] contains_properly
- [ ] is_ccw
- [ ] is_closed
- [ ] is_prepared
- [ ] is_valid_reason

Set operations:
- [ ] coverage_union
- [ ] coverage_union_all
- [ ] intersection_all
- [ ] symmetric_difference_all

Constructive:
- [ ] build_area
- [ ] clip_by_rect (#1928)
- [ ] delaunay_triangles
- [ ] extract_unique_points
- [ ] make_valid
- [ ] minimum_bounding_circle
- [ ] normalize
- [ ] offset_curve
- [ ] polygonize
- [ ] polygonize_full
- [ ] reverse
- [ ] segmentize (requires GEOS 3.10 which is not released yet, so I'd wait with that one)
- [ ] snap
- [ ] voronoi_polygons

Linestring operations:
- [ ] line_interpolate_point
- [ ] line_locate_point
- [ ] line_merge
- [ ] shared_paths
- [ ] shortest_line

Coordinate ops:
- [ ] count_coordinates
- [ ] get_coordinates

**Comment by bitanb1999** [comment]
Hello @martinfleis, I believe for implementing the measurements such as Frechet, Hausdorff distance, and so on, we would have to consider the combinations of the different trajectories formed by the edges of the polygons of the GeoSeries elements or is straight-up between two shapely objects?

**Comment by martinfleis** [comment]
@bitanb1999 No, it is a distance between two geometries. See the [pygeos API](https://pygeos.readthedocs.io/en/stable/measurement.html?highlight=hausdorff#pygeos.measurement.hausdorff_distance). 

We need to turn this API:
```py
pygeos.hausdorff_distance(a, b, densify=None, **kwargs)
```

To this API:

```py
geopandas.GeoSeries.hausdorff_distance(b, densify=None)
```

It is just about exposing those functions as they are as `GeoPandasBase` methods, nothing else. In most of the cases, it should be very straightforward as we just need to mirror how other functions are exposed.

**Comment by bitanb1999** [comment]
@martinfleis Thank you for the clarification! That indeed seems pretty straightforward! Is this issue up for the GSoC 2022?

**Comment by martinfleis** [comment]
> Is this issue up for the GSoC 2022?

Probably not, the small-size GSoC project is 175 hours. Implementing all listed above will be a matter of <10 hours. Have a look at GSoC project ideas here https://github.com/geopandas/geopandas/wiki/Google-Summer-of-Code-2022 to get a sense of scale. Or pick one of them directly :).

**Comment by bitanb1999** [comment]
> > Is this issue up for the GSoC 2022?
> 
> Probably not, the small-size GSoC project is 175 hours. Implementing all listed above will be a matter of <10 hours. Have a look at GSoC project ideas here https://github.com/geopandas/geopandas/wiki/Google-Summer-of-Code-2022 to get a sense of scale. Or pick one of them directly :).

Oh okay! Thank you!

**Comment by bitanb1999** [comment]
> > Is this issue up for the GSoC 2022?
> 
> Probably not, the small-size GSoC project is 175 hours. Implementing all listed above will be a matter of <10 hours. Have a look at GSoC project ideas here https://github.com/geopandas/geopandas/wiki/Google-Summer-of-Code-2022 to get a sense of scale. Or pick one of them directly :).

Another question: Is it possible to integrate GEE with GeoPandas for heavy research-based visualisations?

**Comment by martinfleis** [comment]
> Is it possible to integrate GEE with GeoPandas for heavy research-based visualisations?

I think that `geemap` is your best shot at the moment https://geemap.org/notebooks/geopandas/. But we're out of topic here :).

**Comment by bitanb1999** [comment]
> > Is it possible to integrate GEE with GeoPandas for heavy research-based visualisations?
> 
> I think that `geemap` is your best shot at the moment https://geemap.org/notebooks/geopandas/. But we're out of topic here :).

Yes, the question was associated with the GSoC ideas. Thank you for answering!

**Comment by martinfleis** [comment]
@bitanb1999 If you want to have a chat about GSoC ideas, feel free to move it to out gitter https://gitter.im/geopandas/geopandas. Happy to have a discussion there!

**Comment by EwoutH** [comment]
> Probably not, the small-size GSoC project is 175 hours. Implementing all listed above will be a matter of <10 hours.

Do you still think this estimate is accurate?

It would be very nice to have all those functions exposed, and if this could be a potential GSoC project, the time to start thinking about that is [around now](https://developers.google.com/open-source/gsoc/timeline).

**Comment by bretttully** [comment]
I'm very keen to get the precision functions from shapely2/pygeos in. If someone can point me to the best place to put them, I am happy to have a go at a PR.

Ideally, I'd like to have an API like

```python
gdf = gpd.GeoDataFrame(...)  # maybe an optional ctr param?
gdf.set_precision(...)  # with and without inplace
```

One thing I would **really** like is to be able to (de)serialize this via files like geoparquet, but I suspect that is asking way too much, or at least can be split into a distinct issue :-)

**Comment by martinfleis** [comment]
With `set_precision` and any other function that takes an array of geometries and returns another one. you can roughly follow what has been done in https://github.com/geopandas/geopandas/pull/2621.

**Comment by JamesGardiner** [comment]
I have some time to add a number of these methods (hopefully all but that may be a stretch goal). Would you accept them as separate PRs, or would you prefer one single PR that adds all the shapely functions?

**Comment by martinfleis** [comment]
@JamesGardiner That would be great. I will also have some time to work on this, so it may good to synchronise a bit. 

Make an individual PR per method please, it is easier to review small, well defined contributions.

**Comment by JamesGardiner** [comment]
Great, I'll let you know how I'm getting on. Already started on offset_circle so will get a PR open for that soon

**Comment by martinfleis** [comment]
@JamesGardiner Thanks for the steady flow of PRs! I'll do the reviews during the next week ;).

**Comment by JamesGardiner** [comment]
No rush @martinfleis, each is quite small so won't be difficult to keep them up to date!

**Comment by JamesGardiner** [comment]
Thanks for the reviews @martinfleis - should have time to take a look at those today

**Comment by ljwolf** [comment]
Are things like `shapely.get_parts()` or `shapely.get_rings()` in scope? `get_parts()` would be a more explicit call to `geometry.explode()`, but `get_rings()` would be distinctive.

**Comment by martinfleis** [comment]
`get_parts` is considered exposed as `explode`. Not sure how else would it work if it was explicit? `GeoSeries.explode()` is practically just `get_parts` + index handling. 

`get_rings` could work the same as `explode` I suppose.

**Comment by SebastianoX** [comment]
Quick suggestion here: why not using accessors, as pandas does with datetimes (`DataFrame.dt.<datetime methods>`)?

It would become `GeoSeries.sh.<shapely mehtods>`.

So instead of:
```python
GeoSeries.apply(shapely.concave_hull)
```

It would become:
```python
GeoSeries.sh.concave_hull
```

**Comment by martinfleis** [comment]
Historical reasons. GeoPandas API predates accessor API as far as I remember. It would make more sense if the original object was still a `pandas.DataFrame` with an extension array but we predate that possibility either. With the Geo* subclasses, it feels okay to expose all these directly.

**Comment by EwoutH** [comment]
I would love to be able to use `GeoSeries.contains_xy()` in the future!

**Comment by martinfleis** [comment]
@EwoutH feel free to come up with a PR for that! Otherwise I'll tackle it before 1.0 is out for sure.

**Comment by martinfleis** [comment]
`get_dimensions` has overlapping usage with `has_z` and `is_empty`. Not sure if it is worth implementing.

**Comment by martinfleis** [comment]
`get_num_points`, which returns a number of points in LineString and LinearRing only is also a candidate to ignore, given `get_num_coordinates` provides the same number for any geometry type.
